packages/aws-cdk/lib/api/util/template-body-parameter.ts

```diff
@@ -16,38 +17,44 @@ export type TemplateBodyParameter = {
 
 const LARGE_TEMPLATE_SIZE_KB = 50;
 
+type MakeBodyParameterResult =
+  | { type: 'direct'; param: TemplateBodyParameter }
+  | { type: 'upload'; addToManifest: (builder: AssetManifestBuilder) => TemplateBodyParameter };
+
 /**
  * Prepares the body parameter for +CreateChangeSet+.
  *
- * If the template is small enough to be inlined into the API call, just return
- * it immediately.
- *
- * Otherwise, add it to the asset manifest to get uploaded to the staging
- * bucket and return its coordinates. If there is no staging bucket, an error
- * is thrown.
+ * If the template has already been uploaded to an asset bucket or the template
+ * is small enough to be inlined into the API call, returns a 'direct' type response
+ * with can go into the CloudFormation API call. The `_assetsPublishedProof` parameter
+ * exists to statically prove that `publishAssets` has been called already.
  *
- * @param stack     the synthesized stack that provides the CloudFormation template
- * @param toolkitInfo information about the toolkit stack
+ * Otherwise, returns an object with an `addToManifest` function; call that with an `AssetManifestBuilder`
+ * (and publish the added artifacts!) to obtain the CloudFormation API call parameters.
+ * there is no staging bucket, an error is thrown.
  */
 export async function makeBodyParameter(
   stack: cxapi.CloudFormationStackArtifact,
   resolvedEnvironment: cxapi.Environment,
-  assetManifest: AssetManifestBuilder,
   resources: EnvironmentResources,
   sdk: ISDK,
-  overrideTemplate?: any,
-): Promise<TemplateBodyParameter> {
+  _assetsPublishedProof: AssetsPublishedProof,
```

Comment:
Unused? Don't we need to check this?

Reply:
The fact that you can pass an object that passes the type checker is proof that you have one. 

Yes someone could do `as any`, but we're not trying to protect against someone who is willingly trying to bypass the system, we are protecting against accidental misuse.